bson/flexible_bson/writer.py

```diff
@@ -0,0 +1,288 @@
+# Copyright 2018-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""BSON (Binary JSON) encoding.
+
+"""
+
+from collections import deque
+from enum import Enum
+from io import BytesIO
+
+from bson import _make_name, _name_value_to_bson, _PACK_INT
+from bson.codec_options import DEFAULT_CODEC_OPTIONS
+from bson.objectid import ObjectId
+from bson.py3compat import text_type
+
+
+class BSONWriterException(BaseException):
+    pass
+
+
+class BSONTypes(Enum):
+    X00 = b"\x00"
+    FLOAT = b"\x01"
+    STRING = b"\x02"
+    DOCUMENT = b"\x03"
+    ARRAY = b"\x04"
+    BINARY = b"\x05"
+    UNDEFINED = b"\x06"
+    OBJECT_ID = b"\x07"
+    BOOLEAN = b"\x08"
+    UTCDATETIME = b"\x09"
+    NULL = b"\x0A"
+    INT = b"\x10"
+    TIMESTAMP = b"\x11"
+    LONG = b"\x12"
+    DECIMAL128 = b"\x13"
+
+
+class _BSONWriterState(Enum):
+    INITIAL = 0
+    TOP_LEVEL = 1
+    DOCUMENT = 2
+    ARRAY = 3
+    FINAL = 4
+
+
+_ID_FIELD_C_NAME = _make_name("_id")
+
+
+class BSONWriter(object):
+    def __init__(self, check_keys=False, codec_options=None):
+        self._check_keys = check_keys
+        self._codec_options = codec_options or DEFAULT_CODEC_OPTIONS
+        self.initialize()
+
+    def initialize(self):
+        # Writer state tracker.
+        self._states = deque()
+        self._update_state(_BSONWriterState.INITIAL)
+
+        # Bytestream in which to place BSON encoded bytes.
+        # Deque used to track streams for nested entities.
+        self._streams = deque()
```

Comment:
Can you talk a bit about the advantages of having each subdocument or array be a separate byte stream, instead of using one for the entire document?

Reply:
Yes I'd like to know too. I think it would be more efficient to reserve space for the bson document/array size, save the position, and patch the actual size in when the element is ended.